### PR TITLE
LPD-36494 Convert properties files from the `/language` folder to a batch client extension at build time

### PR DESCRIPTION
**Pull request by thiago-buarqque**
# Goal
To provide the ability of importing translations in PaaS/SaaS using batch client extensions. 

# Approach
Convert `Language*.properties` files to a valid batch engine file (batch-engine-data.json) so it can be easily imported using the cx batch framework. 

As a suggestion from Marco Leo, we are not asking the user to create a client extension-like folder (i.e. adding the client-extension.yaml file). We are expecting the `Language.properties` files to be inside the `${root_folder}/language/lang` folder.

#### `buildLang`

We're also providing the ability to run the `buildLang` task to help copying the translations to different languages.

# Example

Language.properties
```properties
sample-language-key=Sample Language Value
```

Run `gw buildLang`. The output will be 

```bash
├── dist
│   └── language.zip
├── lang
│   ├── Language_ar.properties
│    ...
│   ├── Language_nb.properties
│   ├── Language_nl_BE.properties
│   ├── Language_nl.properties
│   ├── Language_pl.properties
│   ├── Language.properties
│   ├── Language_pt_BR.properties
│   ├── Language_pt_PT.properties
│    ...
│   ├── Language_ro.properties
│   ├── Language_ru.properties

└── README.markdown
```

Run `gw deploy` and you get a zip file in a batch client extension format. The important stuff is `build/liferay-client-extension-build/batch/language-batch-engine-data.json`

```json
{
    "configuration": {
        "className": "com.liferay.portal.language.rest.dto.v1_0.Message",
        "parameters": {
            "containsHeaders": "true",
            "importStrategy": "ON_ERROR_FAIL",
            "updateStrategy": "UPDATE"
        },
        "taskItemDelegateName": "DEFAULT"
    },
    "items": [
        {
            "key": "sample-language-key",
            "languageId": "en_US",
            "value": "Sample Language Value"
        }
    ]
}
```

Notes regarding the generated batch file:

1. The buildLang task does not generate translations for `en_US`, only `en` and other variations of English. But we are safe to consider `en` as `en_US` (see why [here](https://liferay.slack.com/archives/C03LW8NUK/p1726692063600729));
2. Automatic translations, the ones marked with ` (Automatic Copy)` at the end, are ignored;
3. The `buildLang` task also runs when running `gw deploy`.

### CODE REVIEW

**Comment by thiago-buarqque** [comment]
@drewbrokke @ethib137 guys, please review the README file I added to the sample cx, I'm not the best at writing docs :sweat_smile:

**Comment by drewbrokke** [comment]
@thiago-buarqque I think a lot of your PR description is a good enough for the README file... :)

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java [hunk 1]
```diff
@@ -1089,6 +1199,41 @@ private boolean _isActiveProfile(Project project, String profileName) {
 		return false;
 	}
 
+	private boolean _isSpecialLanguageProject(File rootDir, File projectDir) {
```
Comment (drewbrokke) [inline modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java:1202]: This method name needs to change. I think it's best to put it into a dedicated static util under the name `isLanguageProject` or something like that.

Comment (drewbrokke) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java:1202]: Move its overloads, too.

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java [hunk 2]
```diff
@@ -310,6 +309,99 @@ else if (clientExtension.type.equals("themeCSS")) {
 			createClientExtensionConfigTaskProvider, workspaceExtension);
 
 		_configureLiferayRoutes(project, workspaceExtension);
+
+		if (_isSpecialLanguageProject(project)) {
```
Comment (drewbrokke) [inline modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java:313]: This whole block should become its own _configureLanguageProject-style method.

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java [hunk 3]
```diff
@@ -310,6 +309,99 @@ else if (clientExtension.type.equals("themeCSS")) {
...
+			TaskProvider<GenerateLanguageBatchConfigTask>
+				generateLanguageBatchConfigTaskProvider =
+					GradleUtil.addTaskProvider(
+						project, GENERATE_LANGUAGE_BATCH_TASK_NAME,
+						GenerateLanguageBatchConfigTask.class);
```
Comment (drewbrokke) [inline modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java:318]: Generally, for tasks like this we should use a two-step structure of:
1. Call `_createGenerateLanguageBatchConfigTaskProvider` that creates the task and adds any necessary metadata (group, description, etc).
2. Call `_configureGenerateLangaugeBatchConfigTask` that configures the task with dependencies and any inputs.

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java [hunk 4]
```diff
@@ -104,6 +110,9 @@ public class ClientExtensionProjectConfigurator
 	public static final String CREATE_CLIENT_EXTENSION_CONFIG_TASK_NAME =
 		"createClientExtensionConfig";
 
+	public static final String GENERATE_LANGUAGE_BATCH_TASK_NAME =
+		"generateLanguageBatch";
```
Comment (drewbrokke) [inline modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java:114]: This task name should match the task class name. We should go with whichever is the most accurate.

**Review by drewbrokke — CHANGES_REQUESTED** [review]

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java (hunk 1 above)
Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java:1202]: @drewbrokke  Where should I place that static util class? We have `internal/util`, but that feels generic. Maybe create `/internal/client/extension/util` and place it there.

Name suggestion: `LanguageClientExtensionUtil`

Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/configurator/ClientExtensionProjectConfigurator.java:1202]: I put it in  `/internal/client/extension/util`, seems to make more sense

**Comment by thiago-buarqque** [comment]
@drewbrokke changes made, please let me know if there is additional ones to be made

File: modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/client/extension/util/LanguageBatchUtil.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.Objects;
+
+import org.gradle.api.Project;
+
+/**
+ * @author Thiago Buarque
+ */
+public class LanguageBatchUtil {
+
+	public static boolean isSpecialLanguageProject(
```
Comment (ethib137) [inline modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/client/extension/util/LanguageBatchUtil.java:23]: I may have missed more discussion around this, but shouldn't we just call this `isLanguageProject`?

Comment (thiago-buarqque) [reply modules/sdk/gradle-plugins-workspace/src/main/java/com/liferay/gradle/plugins/workspace/internal/client/extension/util/LanguageBatchUtil.java:23]: Ops, you are right. Missed that one

**Comment by drewbrokke** [comment]
@thiago-buarqque see https://github.com/liferay-devtools/liferay-portal/pull/252. Will you please verify that the solution still works as intended on the DXP side of things?

**Pull request closed without merging**